ibis/common/caching.py

```diff
@@ -47,56 +49,50 @@ def __init__(
         generate_name: Callable[[], str],
         key: Callable[[Any], Any],
     ) -> None:
-        self.cache = bidict()
-        # Somehow mypy needs a type hint here
-        self.refs: Counter = Counter()
         self.populate = populate
         self.lookup = lookup
         self.finalize = finalize
-        # Somehow mypy needs a type hint here
-        self.names: defaultdict = defaultdict(generate_name)
+        self.generate_name = generate_name
         self.key = key or (lambda x: x)
 
+        self.cache: dict[Any, CacheEntry] = dict()
+
     def get(self, key, default=None):
-        try:
-            return self[key]
-        except KeyError:
-            return default
+        if (res := self.cache.get(key)) is not None:
+            return res.ref()
+        return default
 
     def __getitem__(self, key):
-        result = self.cache[key]
-        self.refs[key] += 1
-        return result
+        return self.cache[key].ref()
 
-    def store(self, input) -> None:
+    def store(self, input):
         """Compute and store a reference to `key`."""
         key = self.key(input)
-        name = self.names[key]
+        name = self.generate_name()
         self.populate(name, input)
-        self.cache[key] = self.lookup(name)
-        # nothing outside of this instance has referenced this key yet, so the
-        # refcount is zero
-        #
-        # in theory it's possible to call store -> delitem which would raise an
-        # exception, but in practice this doesn't happen because the only call
-        # to store is immediately followed by a call to getitem.
-        self.refs[key] = 0
-
-    def __delitem__(self, key) -> None:
-        # we need to remove the expression representing the computed physical
-        # table as well as the expression that was used to create that table
-        #
-        # bidict automatically handles this for us; without it we'd have to do
-        # to the bookkeeping ourselves with two dicts
-        if (inv_key := self.cache.inverse.get(key)) is None:
-            raise IbisError(
-                "Key has already been released. Did you call "
-                "`.release()` twice on the same expression?"
-            )
-
-        self.refs[inv_key] -= 1
-        assert self.refs[inv_key] >= 0, f"refcount is negative: {self.refs[inv_key]:d}"
-
-        if not self.refs[inv_key]:
-            del self.cache[inv_key], self.refs[inv_key]
-            self.finalize(key)
+        cached = self.lookup(name)
+        finalizer = finalize(cached, self._release, key)
+        finalizer.atexit = False
```

Comment:
Does this mean that cached tables are not guaranteed to be cleaned up?

Reply:
`atexit` means when called by the interpreter shutdown ([docs](https://docs.python.org/3/library/atexit.html)), ie. in a situation where the sessions are being closed (or have been, actually). I assumed that, since we create temporary tables, which afaik are only valid for a current session, it was redundant to explicitly request their removal.

Reply:
Do we want to remove the `DeprecationWarning` and maybe update the docs and release notes to make the new behaviour better known, then?

Reply:
It's not necessarily redundant, because unfortunately not all backends have support for temporary tables.

Reply:
I *think* we want keep the default value of `atexit` (`True`), and allow the cleanup to run on interpreter shutdown.

Reply:
> I think we want keep the default value of atexit (True), and allow the cleanup to run on interpreter shutdown.

Ok, then let's test that.

Reply:
> depends on whether the database connection is still there

True, and that's a big if since it renders the handler useless if the connection isn't there.

I'd say let's leave the default behavior for now.